keras_nlp/samplers/sampler.py

```diff
@@ -358,20 +410,28 @@ def one_step(
                 return current_index, prompt, mask
             return [current_index, prompt, mask, cache]
 
+        def loop_cond(current_index, prompt, mask, cache=None):
+            if end_token_id is None:
+                return tf.less(current_index, max_length)
+            end_token_seen = (prompt == end_token_id) & (
+                original_padding_mask == 0
+            )
+            sequence_done = tf.reduce_any(end_token_seen, axis=-1)
+            all_done = tf.reduce_all(sequence_done)
+            return tf.logical_not(all_done) and tf.less(
+                current_index, max_length
+            )
+
         if cache is None:
             current_index, prompt, mask = tf.while_loop(
-                cond=lambda current_index, prompt, mask: tf.less(
-                    current_index, max_length
-                ),
+                cond=loop_cond,
                 body=one_step,
                 loop_vars=[current_index, prompt, mask],
```

Comment:
One small cleanup we could do is pass `max_length - current_index` as `maximum_iterations` here, and let the `cond` be all about the early stopping condition and not track the index at all.

That seems nice and readable (possibly this could also have compilation benefits, not sure). HF does this https://github.com/huggingface/transformers/blob/main/src/transformers/generation/tf_utils.py#L1726-L1735

Reply:
thx! done